src/architecture/merkledb.md

```diff
@@ -0,0 +1,207 @@
+
+- Add and remove set items
+- Check if an item is already present using the item itself or its hash
+- Iterate over stored items in the lexicographic order of their hashes
+- Iterate over hashes of items in the lexicographic order
+- Clear the set (i.e., remove all items).
+
+The hash used in `ValueSetIndex` is calculated using the `object_hash()` method
+of the [`ObjectHash` trait][object-hash].
+
+!!! summary "Implementation Details"
+    Internally, `ValueSetIndex` uses hashes of items as keys,
+    and items themselves as corresponding values.
+
+### KeySetIndex
+
+[`KeySetIndex`][key-set-index] represents a set.
+The following procedures are implemented:
+
+- Add and remove set items
+- Check if a specific item is in the set
+- Iterate over items in the lexicographic order
+- Clear the set (i.e., remove all stored items).
+
+!!! summary "Implementation Details"
+    Internally, the item is used as a key, and its value is always empty.
+
+#### KeySetIndex vs ValueSetIndex
+
+While `ValueSetIndex` uses a hash as a key, `KeySetIndex` puts an entire binary
+serialization of an item into the key.
+
+- `KeySetIndex` does not have an additional overhead on hashing
+  set items.
+- `KeySetIndex` should not be used when set items are relatively big;
+  only small items should be stored in it (such as integers, small strings,
+  small
+  tuples). On the other hand, `ValueSetIndex` handles storing big and complex
+  items more easily.
+- `KeySetIndex` introduces a lexicographical order over stored
+  items. In `ValueSetIndex` items are ordered according to their hash
+  function properties.
+
+### Entry
+
+`Entry` represents a set that contains only one item.
+
+The following operations are implemented:
+
+- Get, set and remove value
+- Check if the value is present.
+
+### Merkelized Objects
+
+Merkelized objects represent a list and a map with additional
+features. Such objects can create proofs of existence or absence for
+stored data items.
+
+When a light client requests data from an Exonum full node, the proof can be
+built and sent along with the actual data. Having block headers
+and this proof, the client may check that the received data was really
+authorized by the validators without having to replicate the entire blockchain
+contents.
+
+#### ProofListIndex
+
+[`ProofListIndex`][proof-list-index] implements a Merkle
+tree, which is a merkelized version of an
+array list. It implements the same methods as `ListIndex`, and adds an
+additional feature: based on Merkle trees, `ProofListIndex` allows efficiently
+creating compact proofs of existence for the list items.
+The following additional procedures are implemented:
+
+- Get the height of the Merkle tree. As the tree is balanced (though may be not
+  full), its height is close to `log2` of the list length
+- Get the value of the tree root (i.e., the hash of the entire Merkle tree)
+- Build a proof of existence/absence for an item at a specific position
```

Comment:
If I remember correctly, out-of-range proofs couldn't be built. Has this been implemented?

Reply:
Yes, we implemented it, you can now get proof of absence of a single element or range from `ProofListIndex`.

Reply:
That's neat :+1: